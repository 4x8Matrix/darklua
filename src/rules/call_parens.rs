use crate::nodes::{Arguments, Block, Expression, FunctionCall, StringExpression, TableExpression};
use crate::process::{DefaultVisitorMut, NodeProcessorMut, NodeVisitorMut};
use crate::rules::{Rule, RuleConfigurationError, RuleProperties};

use std::mem;

#[derive(Debug, Clone, Default)]
struct Processor {}

impl NodeProcessorMut for Processor {
    fn process_function_call(&mut self, call: &mut FunctionCall) {
        let new_arguments = match call.mutate_arguments() {
            Arguments::Tuple(expressions) if expressions.len() == 1 => {
                let expression = expressions.iter_mut().next().unwrap();

                match expression {
                    Expression::String(string) => {
                        let mut steal_string = StringExpression::empty();
                        mem::swap(string, &mut steal_string);
                        Some(Arguments::String(steal_string))
                    }
                    Expression::Table(table) => {
                        let mut steal_table = TableExpression::default();
                        mem::swap(table, &mut steal_table);
                        Some(Arguments::Table(steal_table))
                    }
                    _ => None,
                }
            }
            _ => None,
        };

        if let Some(new_arguments) = new_arguments {
            *call.mutate_arguments() = new_arguments;
        }
    }
}

pub const REMOVE_FUNCTION_CALL_PARENS: &'static str = "remove_function_call_parens";

/// A rule that removes parentheses when calling functions with a string or a table.
#[derive(Debug, Default, PartialEq, Eq)]
pub struct RemoveFunctionCallParens {}

impl Rule for RemoveFunctionCallParens {
    fn process(&self, block: &mut Block) {
        let mut processor = Processor::default();
        DefaultVisitorMut::visit_block(block, &mut processor);
    }

    fn configure(&mut self, properties: RuleProperties) -> Result<(), RuleConfigurationError> {
        for (key, _value) in properties {
            return Err(RuleConfigurationError::UnexpectedProperty(key))
        }

        Ok(())
    }

    fn get_name(&self) -> &'static str {
        REMOVE_FUNCTION_CALL_PARENS
    }

    fn serialize_to_properties(&self) -> RuleProperties {
        RuleProperties::new()
    }
}

#[cfg(test)]
mod test {
    use super::*;

    use insta::assert_json_snapshot;

    fn new_rule() -> RemoveFunctionCallParens {
        RemoveFunctionCallParens::default()
    }

    #[test]
    fn serialize_default_rule() {
        let rule: Box<dyn Rule> = Box::new(new_rule());

        assert_json_snapshot!("default_remove_function_call_parens", rule);
    }
}
