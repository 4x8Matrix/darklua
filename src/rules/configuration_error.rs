use std::fmt;

/// When implementing the configure method of the Rule trait, the method returns a result that uses
/// this error type.
#[derive(Debug, Clone, Eq, PartialEq)]
pub enum RuleConfigurationError {
    /// When a rule gets an unknown property. The string should be the unknown field name.
    UnexpectedProperty(String),
    /// When a rule has a required property. The string should be the field name.
    MissingProperty(String),
    /// When a property is associated with something else than an expected string. The string is
    /// the property name.
    StringExpected(String),
    /// When a property is associated with something else than an expected unsigned number. The
    /// string is the property name.
    UsizeExpected(String),
    /// When a property is associated with something else than an expected list of strings. The
    /// string is the property name.
    StringListExpected(String),
    /// When the value type is invalid. The string is the property name that was given the wrong
    /// value type.
    UnexpectedValueType(String),
}

impl fmt::Display for RuleConfigurationError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        use RuleConfigurationError::*;

        match self {
            UnexpectedProperty(property) => write!(f, "unexpected field '{}'", property),
            MissingProperty(property) => write!(f, "missing required field '{}'", property),
            StringExpected(property) => write!(f, "string value expected for field '{}'", property),
            UsizeExpected(property) => {
                write!(f, "unsigned integer expected for field '{}'", property)
            }
            StringListExpected(property) => {
                write!(f, "list of string expected for field '{}'", property)
            }
            UnexpectedValueType(property) => write!(f, "unexpected type for field '{}'", property),
        }
    }
}
