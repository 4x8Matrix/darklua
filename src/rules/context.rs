use crate::nodes::Block;

use std::collections::HashMap;

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
enum Feature {
    ShallowMerge,
}

/// The intent of this struct is to hold data shared across all rules applied to a file.
#[derive(Debug, Clone, Default)]
pub struct Context {
    features: HashMap<Feature, String>,
}

impl Context {
    pub fn pre_process_block(&mut self, block: &Block) {

    }

    pub fn post_process_block(&self, block: &mut Block) {

    }

    /// Returns an identifier for a shallow merge function
    pub fn request_shallow_merge_identifier(&mut self) -> String {
        if let Some(identifier) = self.features.get(&Feature::ShallowMerge) {
            identifier.to_owned()
        } else {
            let id = "_DARKLUA_SHALLOW_MERGE".to_owned();
            self.features.insert(Feature::ShallowMerge, id.clone());
            id
        }
    }
}
