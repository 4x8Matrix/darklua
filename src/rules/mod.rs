//! A module that contains the different rules that mutates a Lua block.

mod call_parens;
mod compute_expression;
mod configuration_error;
mod convert_index_to_field;
mod empty_do;
mod group_local;
mod inject_value;
mod method_def;
mod no_local_function;
mod remove_comments;
mod remove_spaces;
mod rename_variables;
mod rule_property;
mod unused_if_branch;
mod unused_while;
mod virtual_execution;

pub use call_parens::*;
pub use compute_expression::*;
pub use configuration_error::RuleConfigurationError;
pub use convert_index_to_field::*;
pub use empty_do::*;
pub use group_local::*;
pub use inject_value::*;
pub use method_def::*;
pub use no_local_function::*;
pub use remove_comments::*;
pub use remove_spaces::*;
pub use rename_variables::*;
pub use rule_property::*;
pub use unused_if_branch::*;
pub use unused_while::*;
pub use virtual_execution::*;

use crate::nodes::Block;

use serde::de::{self, MapAccess, Visitor};
use serde::ser::SerializeMap;
use serde::{Deserialize, Deserializer, Serialize, Serializer};
use std::collections::HashMap;
use std::fmt;
use std::str::FromStr;

/// The intent of this struct is to hold data shared across all rules applied to a file.
#[derive(Debug, Clone, Default)]
pub struct Context {}

pub type RuleProcessResult = Result<(), Vec<String>>;

/// Defines an interface that will be used to mutate blocks and how to serialize and deserialize
/// the rule configuration.
pub trait Rule: RuleConfiguration {
    /// This method should mutate the given block to apply the rule.
    fn process(&self, block: &mut Block, context: &mut Context) -> RuleProcessResult;
}

pub trait RuleConfiguration {
    /// The rule deserializer will construct the default rule and then send the properties through
    /// this method to modify the behavior of the rule.
    fn configure(&mut self, properties: RuleProperties) -> Result<(), RuleConfigurationError>;
    /// This method should return the unique name of the rule.
    fn get_name(&self) -> &'static str;
    /// For implementing the serialize trait on the Rule trait, this method should return all
    /// properties that differs from their default value.
    fn serialize_to_properties(&self) -> RuleProperties;
    /// Returns `true` if the rule has at least one property.
    fn has_properties(&self) -> bool {
        !self.serialize_to_properties().is_empty()
    }
}

pub trait FlawlessRule {
    fn flawless_process(&self, block: &mut Block, context: &mut Context);
}

impl<T: FlawlessRule + RuleConfiguration> Rule for T {
    fn process(&self, block: &mut Block, context: &mut Context) -> RuleProcessResult {
        self.flawless_process(block, context);
        Ok(())
    }
}

/// A function to get the default rule stack for darklua. All the rules here must preserve all the
/// functionalities of the original code after being applied. They must guarantee that the
/// processed block will work as much as the original one.
pub fn get_default_rules() -> Vec<Box<dyn Rule>> {
    vec![
        Box::new(RemoveSpaces::default()),
        Box::new(RemoveComments::default()),
        Box::new(ComputeExpression::default()),
        Box::new(RemoveUnusedIfBranch::default()),
        Box::new(RemoveUnusedWhile::default()),
        Box::new(RemoveEmptyDo::default()),
        Box::new(RemoveMethodDefinition::default()),
        Box::new(ConvertIndexToField::default()),
        Box::new(ConvertLocalFunctionToAssign::default()),
        Box::new(GroupLocalAssignment::default()),
        Box::new(RenameVariables::default()),
        Box::new(RemoveFunctionCallParens::default()),
    ]
}

impl FromStr for Box<dyn Rule> {
    type Err = String;

    fn from_str(string: &str) -> Result<Self, Self::Err> {
        let rule: Box<dyn Rule> = match string {
            VIRTUAL_EXECUTION_RULE_NAME => Box::new(VirtualExecution::default()),
            COMPUTE_EXPRESSIONS_RULE_NAME => Box::new(ComputeExpression::default()),
            CONVERT_INDEX_TO_FIELD_RULE_NAME => Box::new(ConvertIndexToField::default()),
            CONVERT_LOCAL_FUNCTION_TO_ASSIGN_RULE_NAME => {
                Box::new(ConvertLocalFunctionToAssign::default())
            }
            GROUP_LOCAL_ASSIGNMENT => Box::new(GroupLocalAssignment::default()),
            INJECT_GLOBAL_VALUE_RULE_NAME => Box::new(InjectGlobalValue::default()),
            REMOVE_COMMENTS_RULE_NAME => Box::new(RemoveComments::default()),
            REMOVE_EMPTY_DO_RULE_NAME => Box::new(RemoveEmptyDo::default()),
            REMOVE_FUNCTION_CALL_PARENS => Box::new(RemoveFunctionCallParens::default()),
            REMOVE_METHOD_DEFINITION_RULE_NAME => Box::new(RemoveMethodDefinition::default()),
            REMOVE_SPACES_RULE_NAME => Box::new(RemoveSpaces::default()),
            REMOVE_UNUSED_IF_BRANCH_RULE_NAME => Box::new(RemoveUnusedIfBranch::default()),
            REMOVE_UNUSED_WHILE_RULE_NAME => Box::new(RemoveUnusedWhile::default()),
            RENAME_VARIABLES_RULE_NAME => Box::new(RenameVariables::default()),
            _ => return Err(format!("invalid rule name: {}", string)),
        };

        Ok(rule)
    }
}

impl Serialize for Box<dyn Rule> {
    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
        let properties = self.serialize_to_properties();
        let property_count = properties.len();
        let rule_name = self.get_name();

        if property_count == 0 {
            serializer.serialize_str(rule_name)
        } else {
            let mut map = serializer.serialize_map(Some(property_count + 1))?;

            map.serialize_entry("rule", rule_name)?;

            let mut ordered: Vec<(String, RulePropertyValue)> = properties.into_iter().collect();

            ordered.sort_by(|a, b| a.0.cmp(&b.0));

            for (key, value) in ordered {
                map.serialize_entry(&key, &value)?;
            }

            map.end()
        }
    }
}

impl<'de> Deserialize<'de> for Box<dyn Rule> {
    fn deserialize<D: Deserializer<'de>>(deserializer: D) -> Result<Box<dyn Rule>, D::Error> {
        struct StringOrStruct;

        impl<'de> Visitor<'de> for StringOrStruct {
            type Value = Box<dyn Rule>;

            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
                formatter.write_str("rule name or rule object")
            }

            fn visit_str<E>(self, value: &str) -> Result<Self::Value, E>
            where
                E: de::Error,
            {
                let mut rule: Self::Value = FromStr::from_str(value).map_err(de::Error::custom)?;

                rule.configure(RuleProperties::new())
                    .map_err(de::Error::custom)?;

                Ok(rule)
            }

            fn visit_map<M>(self, mut map: M) -> Result<Self::Value, M::Error>
            where
                M: MapAccess<'de>,
            {
                let mut rule_name = None;
                let mut properties = HashMap::new();

                while let Some(key) = map.next_key::<String>()? {
                    match key.as_str() {
                        "rule" => {
                            if rule_name.is_none() {
                                rule_name.replace(map.next_value::<String>()?);
                            } else {
                                return Err(de::Error::duplicate_field("rule"));
                            }
                        }
                        property => {
                            let value = map.next_value::<RulePropertyValue>()?;

                            if properties.insert(property.to_owned(), value).is_some() {
                                return Err(de::Error::custom(format!(
                                    "duplicate field {} in rule object",
                                    property
                                )));
                            }
                        }
                    }
                }

                if let Some(rule_name) = rule_name {
                    let mut rule: Self::Value =
                        FromStr::from_str(&rule_name).map_err(de::Error::custom)?;

                    rule.configure(properties).map_err(de::Error::custom)?;

                    Ok(rule)
                } else {
                    Err(de::Error::missing_field("rule"))
                }
            }
        }

        deserializer.deserialize_any(StringOrStruct)
    }
}

fn verify_no_rule_properties(properties: &RuleProperties) -> Result<(), RuleConfigurationError> {
    if let Some((key, _value)) = properties.iter().next() {
        return Err(RuleConfigurationError::UnexpectedProperty(key.to_owned()));
    }
    Ok(())
}

#[cfg(test)]
mod test {
    use super::*;

    use insta::assert_json_snapshot;

    #[test]
    fn snapshot_default_rules() {
        let rules = get_default_rules();

        assert_json_snapshot!("default_rules", rules);
    }

    #[test]
    fn verify_no_rule_properties_is_ok_when_empty() {
        let empty_properties = RuleProperties::default();

        assert_eq!(verify_no_rule_properties(&empty_properties), Ok(()));
    }

    #[test]
    fn verify_no_rule_properties_is_unexpected_rule_err() {
        let mut properties = RuleProperties::default();
        let some_rule_name = "rule name";
        properties.insert(some_rule_name.to_owned(), RulePropertyValue::None);

        assert_eq!(
            verify_no_rule_properties(&properties),
            Err(RuleConfigurationError::UnexpectedProperty(
                some_rule_name.to_owned()
            ))
        );
    }
}
