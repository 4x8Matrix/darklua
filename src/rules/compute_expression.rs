use crate::nodes::{Block, Expression};
use crate::process::{DefaultVisitorMut, Evaluator, NodeProcessorMut, NodeVisitorMut};
use crate::rules::{Rule, RuleConfigurationError, RuleProperties};

#[derive(Debug, Clone, Default)]
struct Computer {
    evaluator: Evaluator,
}

impl Computer {
    fn replace_with(&self, expression: &mut Expression) -> Option<Expression> {
        match expression {
            Expression::Unary(_) | Expression::Binary(_) => {
                if !self.evaluator.has_side_effects(&expression) {
                    self.evaluator.evaluate(&expression)
                        .to_expression()
                } else {
                    None
                }
            }
            _ => None,
        }
    }
}

impl NodeProcessorMut for Computer {
    fn process_expression(&mut self, expression: &mut Expression) {
        if let Some(replace_with) = self.replace_with(expression) {
            *expression = replace_with;
        }
    }
}

pub const COMPUTE_EXPRESSIONS_RULE_NAME: &'static str = "compute_expression";

/// A rule that compute expressions that do not have any side-effects.
#[derive(Debug, Default, PartialEq, Eq)]
pub struct ComputeExpression {}

impl Rule for ComputeExpression {
    fn process(&self, block: &mut Block) {
        let mut processor = Computer::default();
        DefaultVisitorMut::visit_block(block, &mut processor);
    }

    fn configure(&mut self, properties: RuleProperties) -> Result<(), RuleConfigurationError> {
        for (key, _value) in properties {
            return Err(RuleConfigurationError::UnexpectedProperty(key))
        }

        Ok(())
    }

    fn get_name(&self) -> &'static str {
        COMPUTE_EXPRESSIONS_RULE_NAME
    }

    fn serialize_to_properties(&self) -> RuleProperties {
        RuleProperties::new()
    }
}

#[cfg(test)]
mod test {
    use super::*;

    use insta::assert_json_snapshot;

    fn new_rule() -> ComputeExpression {
        ComputeExpression::default()
    }

    #[test]
    fn serialize_default_rule() {
        let rule: Box<dyn Rule> = Box::new(new_rule());

        assert_json_snapshot!("default_compute_expression", rule);
    }
}
