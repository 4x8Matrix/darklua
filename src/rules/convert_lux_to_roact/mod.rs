use crate::nodes::{
    Arguments,
    Block,
    Expression,
    FieldExpression,
    FunctionCall,
    Prefix,
    LocalFunctionStatement,
    StringExpression,
    TableExpression,
    TableEntry,
    LUXAttribute,
    LUXAttributeName,
    LUXAttributeValue,
    LUXChild,
    LUXExpression,
    LUXElement,
    LUXElementName,
    LUXFragment,
};
use crate::process::{
    NodeProcessor,
    NodeVisitor,
    ScopeVisitor,
    Scope,
};
use crate::rules::{Rule, RuleConfigurationError, RuleProperties};

use std::collections::HashSet;

#[derive(Debug, Clone)]
struct RoactCodeGenerator {
    roact_identifier: String,
    identifiers: Vec<HashSet<String>>,
}

impl RoactCodeGenerator {
    fn is_identifier_used(&self, identifier: &str) -> bool {
        self.identifiers.iter()
            .any(|set| set.contains(identifier))
    }

    fn insert_identifier(&mut self, identifier: &str) {
        if let Some(set) = self.identifiers.last_mut() {
            set.insert(identifier.to_owned());
        } else {
            let mut set = HashSet::new();
            set.insert(identifier.to_owned());
            self.identifiers.push(set);
        }
    }

    fn roact_prefix(&self) -> Prefix {
        Prefix::from_name(&self.roact_identifier)
    }

    fn convert_lux_expression(&self, lux: &LUXExpression) -> Expression {
        match lux {
            LUXExpression::LUXElement(element) => self.convert_element(element).0,
            LUXExpression::LUXFragment(fragment) => self.convert_fragment(fragment),
        }
    }

    fn convert_element(&self, element: &LUXElement) -> (Expression, Option<Expression>) {
        let component_name = match element.get_name() {
            LUXElementName::Identifier(name) => {
                if self.is_identifier_used(name) {
                    Expression::Identifier(name.to_owned())
                } else {
                    StringExpression::from_value(name).into()
                }
            }
            LUXElementName::NamespacedName(_name) => {
                unimplemented!()
            }
            LUXElementName::Members(_root, _members) => {
                unimplemented!()
            }
        };

        let (props, key) = self.convert_attributes(element.get_attributes());

        let empty_props = match &props {
            Expression::Nil => true,
            Expression::Table(table) => table.is_empty(),
            _ => false,
        };

        let mut arguments = Arguments::default()
            .append_argument(component_name);

        if !empty_props {
            arguments.push_argument(props);
        }

        let expression = FunctionCall::new(
            FieldExpression::new(self.roact_prefix(), "createElement").into(),
            arguments,
            None,
        ).into();

        (expression, key)
    }

    fn convert_fragment(&self, fragment: &LUXFragment) -> Expression {
        FunctionCall::new(
            FieldExpression::new(self.roact_prefix(), "createFragment").into(),
            Arguments::default()
                .append_argument(self.convert_children(fragment.get_children())),
            None,
        ).into()
    }

    fn convert_children(&self, children: &Vec<LUXChild>) -> Expression {
        let mut current_children = Vec::new();
        let mut merge_lists = Vec::new();

        for child in children.iter() {
            match child {
                LUXChild::LUXElement(element) => {
                    let (expression, key) = self.convert_element(element);
                    let entry = if let Some(key) = key {
                        TableEntry::from_dictionary_entry(key, expression)
                    } else {
                        expression.into()
                    };
                    current_children.push(entry);
                }
                LUXChild::LUXFragment(fragment) => {
                    let expression = self.convert_fragment(fragment);
                    current_children.push(expression.into());
                }
                LUXChild::Expression(expression) => {
                    if let Some(expression) = expression {
                        current_children.push(expression.clone().into());
                    }
                }
                LUXChild::ExpandedExpression(expression) => {
                    let entries = current_children.drain(..)
                        .collect();
                    merge_lists.push(TableExpression::new(entries).into());
                    merge_lists.push(expression.clone());
                }
            }
        }

        if merge_lists.is_empty() {
            TableExpression::new(current_children).into()
        } else {
            unimplemented!()
        }
    }

    fn convert_attributes(&self, attributes: &Vec<LUXAttribute>) -> (Expression, Option<Expression>) {
        let mut current_props = Vec::new();
        let mut merge_lists = Vec::new();
        let mut key = None;

        for attribute in attributes.iter() {
            match attribute {
                LUXAttribute::Named(attribute) => {
                    let mut value = Some(if let Some(attribute_value) = attribute.get_value() {
                        match attribute_value {
                            LUXAttributeValue::DoubleQuoteString(string)
                            | LUXAttributeValue::SingleQuoteString(string) => {
                                StringExpression::from_value(string).into()
                            }
                            LUXAttributeValue::LuaExpression(expression) => {
                                expression.clone()
                            }
                            LUXAttributeValue::LUXElement(element) => {
                                self.convert_element(element).0
                            }
                            LUXAttributeValue::LUXFragment(fragment) => {
                                self.convert_fragment(fragment)
                            }
                        }
                    } else {
                        Expression::True
                    });

                    let prop: Option<Expression> = match attribute.get_name() {
                        LUXAttributeName::Identifier(identifier) => {
                            Some(StringExpression::from_value(identifier).into())
                        }
                        LUXAttributeName::NamespacedName(namespaced_name) => {
                            match namespaced_name.get_namespace().as_ref() {
                                "event" => {
                                    Some(FieldExpression::new(
                                        FieldExpression::new(
                                            self.roact_prefix(),
                                            "Event",
                                        ).into(),
                                        namespaced_name.get_member()
                                    ).into())
                                }
                                "changed" => {
                                    Some(FieldExpression::new(
                                        FieldExpression::new(
                                            self.roact_prefix(),
                                            "Changed",
                                        ).into(),
                                        namespaced_name.get_member()
                                    ).into())
                                }
                                "roact" => {
                                    match namespaced_name.get_member().as_ref() {
                                        "key" => {
                                            key = value.take();
                                            None
                                        }
                                        "ref" => {
                                            Some(FieldExpression::new(
                                                self.roact_prefix(),
                                                "Ref",
                                            ).into())
                                        }
                                        _ => {
                                            unimplemented!()
                                        }
                                    }
                                }
                                _ => {
                                    unimplemented!()
                                }
                            }
                        }
                    };

                    if let Some((prop, value)) = prop.zip(value) {
                        current_props.push(TableEntry::from_dictionary_entry(
                            prop,
                            value,
                        ));
                    }
                }
                LUXAttribute::Spread(expression) => {
                    let entries: Vec<_> = current_props.drain(..)
                        .collect();
                    if !entries.is_empty() {
                        merge_lists.push(TableExpression::new(entries).into());
                    }
                    merge_lists.push(expression.clone());
                }
            }
        }

        if merge_lists.is_empty() {
            (TableExpression::new(current_props).into(), key)
        } else {
            if current_props.is_empty() {
                if merge_lists.len() == 1 {
                    (merge_lists.pop().unwrap(), key)
                } else {
                    unimplemented!()
                }
            } else {
                unimplemented!()
            }
        }
    }
}

impl Default for RoactCodeGenerator {
    fn default() -> Self {
        Self {
            roact_identifier: "Roact".to_owned(),
            identifiers: Vec::new(),
        }
    }
}

impl Scope for RoactCodeGenerator {
    fn push(&mut self) {
        self.identifiers.push(HashSet::new())
    }

    fn pop(&mut self) {
        self.identifiers.pop();
    }

    fn insert(&mut self, identifier: &mut String) {
        self.insert_identifier(identifier);
    }

    fn insert_local(&mut self, identifier: &mut String, _value: Option<&mut Expression>) {
        self.insert_identifier(identifier);
    }

    fn insert_local_function(&mut self, function: &mut LocalFunctionStatement) {
        self.insert_identifier(function.get_name());
    }
}

impl NodeProcessor for RoactCodeGenerator {
    fn process_expression(&mut self, expression: &mut Expression) {
        match expression {
            Expression::LUX(lux_expression) => {
                let new_expression = self.convert_lux_expression(lux_expression);
                *expression = new_expression;
            }
            _ => {}
        }
    }
}

pub const CONVERT_LUX_TO_ROACT_CODE_RULE_NAME: &'static str = "convert_lux_to_roact_code";

/// A rule that removes LUX expressions and generates code for Roact.
#[derive(Debug, Default, PartialEq, Eq)]
pub struct ConvertLUXToRoactCode {}

impl Rule for ConvertLUXToRoactCode {
    fn process(&self, block: &mut Block) {
        let mut processor = RoactCodeGenerator::default();
        ScopeVisitor::visit_block(block, &mut processor);
    }

    fn configure(&mut self, properties: RuleProperties) -> Result<(), RuleConfigurationError> {
        // TODO: make the roact identifier configurable
        for (key, _value) in properties {
            return Err(RuleConfigurationError::UnexpectedProperty(key))
        }

        Ok(())
    }

    fn get_name(&self) -> &'static str {
        CONVERT_LUX_TO_ROACT_CODE_RULE_NAME
    }

    fn serialize_to_properties(&self) -> RuleProperties {
        RuleProperties::new()
    }
}

// #[cfg(test)]
// mod test {
//     use super::*;

//     use insta::assert_json_snapshot;

//     fn new_rule() -> ConvertLUXToRoactCode {
//         ConvertLUXToRoactCode::default()
//     }

//     #[test]
//     fn serialize_default_rule() {
//         let rule: Box<dyn Rule> = Box::new(new_rule());

//         assert_json_snapshot!("default_convert_lux_to_roact_code", rule);
//     }
// }
