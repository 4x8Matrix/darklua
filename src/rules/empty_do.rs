use crate::nodes::{Block, Statement};
use crate::process::{DefaultVisitorMut, NodeProcessorMut, NodeVisitorMut};
use crate::rules::{Rule, RuleConfigurationError, RuleProperties};

struct EmptyDoFilter {
    mutated: bool,
}

impl EmptyDoFilter {
    pub fn has_mutated(&self) -> bool {
        self.mutated
    }
}

impl Default for EmptyDoFilter {
    fn default() -> Self {
        Self {
            mutated: false,
        }
    }
}

impl NodeProcessorMut for EmptyDoFilter {
    fn process_block(&mut self, block: &mut Block) {
        block.filter_statements(|statement| match statement {
            Statement::Do(do_statement) => {
                self.mutated = do_statement.get_block().is_empty();
                !self.mutated
            },
            _ => true,
        });
    }
}

pub const REMOVE_EMPTY_DO_RULE_NAME: &'static str = "remove_empty_do";

/// A rule that removes empty do statements.
#[derive(Debug, Default, PartialEq, Eq)]
pub struct RemoveEmptyDo {}

impl Rule for RemoveEmptyDo {
    fn process(&self, block: &mut Block) {
        while {
            let mut processor = EmptyDoFilter::default();
            DefaultVisitorMut::visit_block(block, &mut processor);
            processor.has_mutated()
        } {}
    }

    fn configure(&mut self, properties: RuleProperties) -> Result<(), RuleConfigurationError> {
        for (key, _value) in properties {
            return Err(RuleConfigurationError::UnexpectedProperty(key))
        }

        Ok(())
    }

    fn get_name(&self) -> &'static str {
        REMOVE_EMPTY_DO_RULE_NAME
    }

    fn serialize_to_properties(&self) -> RuleProperties {
        RuleProperties::new()
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use crate::nodes::DoStatement;

    use insta::assert_json_snapshot;

    fn new_rule() -> RemoveEmptyDo {
        RemoveEmptyDo::default()
    }

    #[test]
    fn remove_empty_do_statement() {
        let rule = new_rule();

        let mut block = Block::default().with_statement(DoStatement::new(Block::default()));

        rule.process(&mut block);

        assert_eq!(block, Block::default());
    }

    #[test]
    fn remove_nested_empty_do_statement() {
        let rule = new_rule();

        let block_with_do_statement = Block::default()
            .with_statement(DoStatement::default());
        let mut block = Block::default()
            .with_statement(DoStatement::new(block_with_do_statement));

        rule.process(&mut block);

        assert_eq!(block, Block::default());
    }

    #[test]
    fn serialize_default_rule() {
        let rule: Box<dyn Rule> = Box::new(new_rule());

        assert_json_snapshot!("default_remove_empty_do", rule);
    }
}
