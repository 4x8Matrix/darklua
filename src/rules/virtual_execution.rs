use crate::nodes::Block;
use crate::process::engine_impl::{
    create_roblox_math_library, create_tonumber, create_tostring, create_type, create_roblox_string_library,
};
use crate::process::{LuaValue, VirtualLuaExecution};
use crate::rules::{
    Context, FlawlessRule, RuleConfiguration, RuleConfigurationError, RuleProperties,
};

use super::RulePropertyValue;

pub const VIRTUAL_EXECUTION_RULE_NAME: &str = "virtual_execution";

#[derive(Debug, PartialEq, Eq)]
struct EngineGlobal {
    identifier: &'static str,
    create_value: fn() -> LuaValue,
    property_name: &'static str,
}

/// A rule that runs Lua code as much as possible statically.
#[derive(Debug, Default, PartialEq, Eq)]
pub struct VirtualExecution {
    globals: Vec<EngineGlobal>,
}

impl VirtualExecution {
    fn include_globals(&mut self, list: Vec<String>) -> Result<(), RuleConfigurationError> {
        for value in list {
            match value.as_str() {
                "roblox-math" => {
                    self.globals.push(EngineGlobal {
                        identifier: "math",
                        create_value: create_roblox_math_library,
                        property_name: "roblox-math",
                    });
                }
                "roblox-string" => {
                    self.globals.push(EngineGlobal {
                        identifier: "string",
                        create_value: create_roblox_string_library,
                        property_name: "roblox-string",
                    });
                }
                "tonumber" => {
                    self.globals.push(EngineGlobal {
                        identifier: "tonumber",
                        create_value: create_tonumber,
                        property_name: "tonumber",
                    });
                }
                "tostring" => {
                    self.globals.push(EngineGlobal {
                        identifier: "tostring",
                        create_value: create_tostring,
                        property_name: "tostring",
                    });
                }
                "type" => {
                    self.globals.push(EngineGlobal {
                        identifier: "type",
                        create_value: create_type,
                        property_name: "type",
                    });
                }
                _ => {
                    return Err(RuleConfigurationError::StringExpected(format!(
                        "invalid engine globals set `{}`",
                        value
                    )))
                }
            }
        }

        Ok(())
    }
}

impl FlawlessRule for VirtualExecution {
    fn flawless_process(&self, block: &mut Block, _: &mut Context) {
        let mut virtual_execution = self.globals.iter().fold(
            VirtualLuaExecution::default().perform_mutations(),
            |execution, global| {
                execution.with_global_value(global.identifier, (global.create_value)())
            },
        );

        virtual_execution.evaluate_chunk(block);
    }
}

impl RuleConfiguration for VirtualExecution {
    fn configure(&mut self, properties: RuleProperties) -> Result<(), RuleConfigurationError> {
        for (key, value) in properties {
            match key.as_str() {
                "includes" => match value {
                    RulePropertyValue::StringList(includes) => self.include_globals(includes)?,
                    _ => return Err(RuleConfigurationError::StringListExpected(key)),
                },
                _ => return Err(RuleConfigurationError::UnexpectedProperty(key)),
            }
        }

        Ok(())
    }

    fn get_name(&self) -> &'static str {
        VIRTUAL_EXECUTION_RULE_NAME
    }

    fn serialize_to_properties(&self) -> RuleProperties {
        let mut properties = RuleProperties::new();

        if !self.globals.is_empty() {
            properties.insert(
                "includes".to_owned(),
                self.globals
                    .iter()
                    .map(|global| global.property_name.to_owned())
                    .collect(),
            );
        }

        properties
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use crate::rules::Rule;

    use insta::assert_json_snapshot;

    fn new_rule() -> VirtualExecution {
        VirtualExecution::default()
    }

    #[test]
    fn serialize_default_rule() {
        let rule: Box<dyn Rule> = Box::new(new_rule());

        assert_json_snapshot!("default_virtual_execution", rule);
    }
}
