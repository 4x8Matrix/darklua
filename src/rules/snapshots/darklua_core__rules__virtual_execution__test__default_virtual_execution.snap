---
source: src/rules/virtual_execution.rs
expression: rule

---
"virtual_execution"
